### PR TITLE
chore: upate ckb-ics deps

### PR DESCRIPTION
**Pull request by ashuralyk**
p.s. post here just to generate contracts binary and update PR comments once finishing the refactoring

change log:
1. rename ics-conn to ics-connection and ics-chan to ics-channel
2. collect same procedures in ics-connection and ics-channel into a base crate
3. upgrade the abstraction of connection, channel and packet contract logic
4. fix a bug of checking port_id in channel_open_init message

### CODE REVIEW

**Comment by ashuralyk** [comment]
@Flouse refactor is completed except the debug in Forcerelay integration-test process, but it's ok to move to the review stage

File: contracts/ics/base/src/axon_client.rs [hunk 1]
```diff
@@ -1,5 +1,3 @@
-#![no_std]
-
 extern crate alloc;
```
Comment (chaoticlonghair) [inline contracts/ics/base/src/axon_client.rs]: Move `extern crate alloc;` to `lib.rs`.

**Review by chaoticlonghair — CHANGES_REQUESTED** [review]

File: contracts/ics/base/src/axon_client.rs (hunk 1 above)
Comment (blckngm) [reply contracts/ics/base/src/axon_client.rs]: It's in lib.rs.

Comment (chaoticlonghair) [reply contracts/ics/base/src/axon_client.rs]: > It's in lib.rs.

My mistake! I updated my original comment.

Comment (chaoticlonghair) [reply contracts/ics/base/src/axon_client.rs]: Then, `extern crate alloc;` should be moved to `lib.rs`.

File: contracts/ics/base/Cargo.toml [hunk 2]
```diff
@@ -1,16 +1,18 @@
...
-rlp ={ version = "0.5.2", default-features = false}
+ckb-std = "0.13.0"
+ckb-ics-axon = { git = "https://github.com/synapseweb3/ckb-ics.git", rev = "ef124f9" }
+axon-tools-riscv = { version = "0.1.1", features = ["proof"] }
+axon-types = { git = "https://github.com/axonweb3/axon-contract", rev = "8c2338a" }
+rlp = { version = "0.5.2", default-features = false }
+tiny-keccak = { version = "2.0.2", features = ["keccak"] }
 molecule = { version = "0.7", default-features = false }
 
 [features]
```
Comment (chaoticlonghair) [inline contracts/ics/base/Cargo.toml]: `ics-base` has `default = ["debugging"]`.

File: contracts/ics/connection/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+name = "ibc-ckb_contracts-ics-connection"
+version = "0.1.0-alpha.0"
+edition = "2021"
+license = "MIT"
+homepage = "https://github.com/synapseweb3/ibc-ckb-contracts"
+repository = "https://github.com/synapseweb3/ibc-ckb-contracts"
+
+[dependencies]
+ckb-std = "0.13.0"
+ics-base = { version = "*", path = "../base" }
```
Comment (chaoticlonghair) [inline contracts/ics/connection/Cargo.toml]: Use `ics-base` with default features.

File: contracts/ics/connection/Cargo.toml [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+[features]
+default = []
+debugging = ["ics-base/debugging"]
```
Comment (chaoticlonghair) [inline contracts/ics/connection/Cargo.toml]: `debugging = ["ics-base/debugging"]` won't control the feature of `ics-base`.

**Review by chaoticlonghair — DISMISSED** [review]

File: contracts/ics/base/Cargo.toml (hunk 2 above)
Comment (ashuralyk) [reply contracts/ics/base/Cargo.toml]: I think `debugging` is a normal feature flag for ics-base, using it will trigger compiling with debug mode?

Comment (blckngm) [reply contracts/ics/base/Cargo.toml]: Why do you need a debugging feature? Just use https://docs.rs/ckb-std/latest/ckb_std/macro.debug.html

Comment (chaoticlonghair) [reply contracts/ics/base/Cargo.toml]: > Why do you need a debugging feature?

TL;DR: It's a historical legacy issue.

It can be changed, today.

<details><summary>Click for details.</summary>

- This contract copied the first contract in this repository, and that contract did that.

- The first contract started at last year.
  - `ckb-std` was not released from August 2022 to March 2023.
  - The maintainer of the last 3 release of `ckb-std` in August 2022, was left the organization; no maintainer until March 2023.
  In that time, the best solution for avoiding the development to be blocked, was to fork it or write code by selves.

- What's the difference between `ckb_std::debug(..)` and ours?

  - No warnings.
    - You can try `ckb-std = 0.10.0` (the latest version before March 2023).

  - More compatible.
    - If your macro call was too long, so that it was broken into multi-lines, and your formatter added a comma after the last argument, `ckb_std::debug(..)` will throw an error.

  - Can be used in release mode simply.

    - Debug mode: size `+10000%` (approximate).
    - Enable `debug-assertions` in release mode: size `+30%` (approximate), cycles `+20%` (approximate).
    
    Use a feature to control that, it still has chance to deploy the with-logs version into testnet (due to size limit).

</details>

Comment (chaoticlonghair) [reply contracts/ics/base/Cargo.toml]: > I think `debugging` is a normal feature flag for ics-base, using it will trigger compiling with debug mode?

I didn't talk about "trigger compiling with debug mode" or other things.
It's not about what you want to do.

No matter what's your purposes, your code is incorrect.

- `ics-base` has `default = ["debugging"]`.
- The crate which uses `ics-base` with:
  - `ics-base = { version = "*", path = "../base" }`
    That means `ics-base/debugging` will always be enabled.
  - Then, `debugging = ["ics-base/debugging"]` in the crate is useless.

You can write some minimal examples to test, for understanding how features works.

**Comment by ashuralyk** [comment]
@Flouse debug in forcerelay has completed

**Review by blckngm — APPROVED** [review]

File: contracts/ics/packet/src/entry.rs [hunk 5]
```diff
@@ -1,229 +1,13 @@
...
-    let metadata_type_script = hl::load_cell_type(0, Source::CellDep)
-        .map_err(|_| Error::FailedToLoadClientTypeScript)?
-        .unwrap();
-    let client_data = metadata_type_script.args().raw_data();
-    let client_id: [u8; 32] = client_data
-        .as_ref()
-        .try_into()
-        .map_err(|_| Error::FailedToLoadClientId)?;
-    Client::new(client_id, &metadata).map_err(|_| Error::FailedToCreateClient)
+    let client = load_client().unwrap_or_default();
```
Comment (Flouse) [inline contracts/ics/packet/src/entry.rs]: Not sure if we can use `unwrap_or_default` here.
It is generally recommended to avoid using `unwrap_or_default`.
Can we handle the absence of a client in a more controlled and explicit way.

**Review by jjyr — APPROVED** [review]

File: contracts/ics/packet/src/entry.rs (hunk 5 above)
Comment (ashuralyk) [reply contracts/ics/packet/src/entry.rs]: it can be eliminated now cause I adjusted the error handling process after I wrote it

**Review by Flouse — APPROVED** [review]
